### PR TITLE
Restrict the maximum number of base classes aka mixins #264

### LINKED ISSUE #264: Restrict the maximum number of base classes aka mixins
# Rule request

## Thesis

There are use cases that promote putting too many bases classes for your class:

```python
class SnippetList(mixins.ListModelMixin,
                  mixins.CreateModelMixin,
                  mixins.DetailModelMixin,
                  mixins.DeleteModelMixin,
                  generics.GenericAPIView):
```

We need to restrict classes with too many base classes as too complex ones.
We also need to provide an option with reasonable default to make this rule configurable.

I guess `3` should be the default value. But I am open to any other suggestions.

## Reasoning

It is almost never possible to navigate to the desired method of a parent class when you need it with multiple mixins.

It is hard to understand `mro` and `super` calls. Do not overuse this technique.

### PR DESCRIPTION
**Pull request by geoc0ld**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

Closes #264

### CODE REVIEW

File: tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import pytest
+
+from wemake_python_styleguide.violations.complexity import (
+    TooManyBaseClassesViolation,
+)
+from wemake_python_styleguide.visitors.ast.classes import WrongClassVisitor
+
+correct_count = """
+   class CorrectClassName(FirstParentClass,
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py]: Please, remove the tabulation.

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py]: We also use different type of `(...)`.

It should be like so:

```python
class Some(
     Parent1,
     Parent2,
):
```

File: tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    tree = parse_ast_tree(code)
+
+    visitor = WrongClassVisitor(default_options, tree=tree)
+    visitor.run()
+
+    assert_errors(visitor, [])
+
+
+too_many_count = """
+   class SomeClassName(FirstParentClass,
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py]: The same about tabulation and `()`

File: wemake_python_styleguide/options/config.py [hunk 3]
```diff
@@ -181,6 +181,12 @@ class Configuration(object):
             'Maximum number of conditions in a `if` or `while` node.',
         ),
 
+        _Option(
+            '--max-classes-number',
```
Comment (sobolevn) [inline wemake_python_styleguide/options/config.py]: It should be `--max-base-classes`

File: wemake_python_styleguide/options/config.py [hunk 4]
```diff
@@ -181,6 +181,12 @@ class Configuration(object):
...
+            defaults.MAX_CLASSES_NUMBER,
```
Comment (sobolevn) [inline wemake_python_styleguide/options/config.py]: Should be `MAX_BASE_CLASSES`.

File: wemake_python_styleguide/violations/complexity.py [hunk 5]
```diff
@@ -507,3 +509,39 @@ class TooManyForsInComprehensionViolation(ASTViolation):
...
+
+
+class TooManyBaseClassesViolation(ASTViolation):
+    """
+    Restrict the maximum number of base classes.
+
+    Reasoning:
+        It is almost never possible to navigate
+        to the desired method of a parent class
+        when you need it with multiple mixin.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/complexity.py]: `mixins`

File: wemake_python_styleguide/violations/complexity.py [hunk 6]
```diff
@@ -507,3 +509,39 @@ class TooManyForsInComprehensionViolation(ASTViolation):
...
+        It is hard to understand mro and super calls.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/complexity.py]: `mro` and `super` should be wrapped with double ``

File: wemake_python_styleguide/violations/complexity.py [hunk 7]
```diff
@@ -507,3 +509,39 @@ class TooManyForsInComprehensionViolation(ASTViolation):
...
+    Solution:
+        Restrict the number of base classes.
+
+    Example::
+
+       # Correct:
+       class SomeClassName(FirstParentClass, SecondParentClass): ...
+
+       # Wrong:
+       class SomeClassName(FirstParentClass,
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/complexity.py]: The same about `()` applies here

File: tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py [hunk 8]
```diff
@@ -8,9 +8,11 @@
 from wemake_python_styleguide.visitors.ast.classes import WrongClassVisitor
 
 correct_count = """
-   class CorrectClassName(FirstParentClass,
-              SecondParentClass,
-              ThirdParentClass): ...
+class CorrectClassName(
+    FirstParentClass,
+    SecondParentClass,
+    ThirdParentClass
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py]: Missing a trailing comma

File: tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py [hunk 9]
```diff
@@ -30,11 +32,13 @@ def test_correct_count(
...
-              SecondParentClass,
-              ThirdParentClass,
-              CustomClass,
-              AddedClass): ...
+class SomeClassName(
+    FirstParentClass,
+    SecondParentClass,
+    ThirdParentClass,
+    CustomClass,
+    AddedClass
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_counts/test_bases_classes_counts.py]: Missing a trailing comma

File: wemake_python_styleguide/violations/complexity.py [hunk 10]
```diff
@@ -518,8 +518,8 @@ class TooManyBaseClassesViolation(ASTViolation):
     Reasoning:
         It is almost never possible to navigate
         to the desired method of a parent class
-        when you need it with multiple mixin.
-        It is hard to understand mro and super calls.
+        when you need it with multiple mixins.
+        It is hard to understand 'mro' and 'super' calls.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/complexity.py]: Please, use double `` to wrap `mro` and `super`. Like here: https://github.com/wemake-services/wemake-python-styleguide/blob/ffc72dc3fe0a04e36527f9e06ab9b1eccfff4d92/wemake_python_styleguide/violations/complexity.py#L169

**Comment by geoc0ld** [comment]
@sobolevn Used it personally. I am new in python, and this is good opportunity to learn it.

**Pull request merged**